src/tools/vfs/vfs_main.c

```diff
@@ -10,9 +10,370 @@
+            devnull_fd = open("/dev/null", O_WRONLY);
+            dup2(devnull_fd, 1);
+            dup2(devnull_fd, 2);
+            close(devnull_fd);
+        }
+        execvp(argv[0], argv);
+        vfs_error("failed to execute '%s': %m", command);
+        _exit(1);
+    }
+
+    ret = waitpid(child_pid, &status, 0);
+    if (ret < 0) {
+        vfs_error("waitpid(%d) failed: %m", child_pid);
+        return -errno;
+    } else if (WIFEXITED(status) && (WEXITSTATUS(status) != 0)) {
+        vfs_error("'%s' exited with status %d", command, WEXITSTATUS(status));
+        return -1;
+    } else if (!WIFEXITED(status)) {
+        vfs_error("'%s' did not exit properly (%d)", command, status);
+        return -1;
+    }
+
+    return 0;
+}
+
+static int vfs_unlink_socket(int silent_notexist)
+{
+    int ret;
+
+    vfs_log("removing existing socket '%s'", g_sockaddr.sun_path);
+
+    ret = unlink(g_sockaddr.sun_path);
+    if (ret < 0) {
+        ret = -errno;
+        if (silent_notexist && (errno == ENOENT)) {
+            vfs_log("could not unlink '%s': %m", g_sockaddr.sun_path);
+        } else {
+            vfs_error("could not unlink '%s': %m", g_sockaddr.sun_path);
+        }
+        return ret;
+    }
+
+    return 0;
+}
+
+/* return 0 or the (negative) value of errno in case of error */
+static int vfs_listen(int silent_addinuse_err)
+{
+    int listen_fd, ret;
+
+    ret = umask(~S_IRWXU);
+    if (ret < 0) {
+        ret = -errno;
+        vfs_error("failed to set umask permissions: %m");
+        goto out;
+    }
+
+    listen_fd = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (listen_fd < 0) {
+        ret = -errno;
+        vfs_error("failed to create listening socket: %m");
+        goto out;
+    }
+
+    ret = bind(listen_fd, (const struct sockaddr*)&g_sockaddr,
+               sizeof(g_sockaddr));
+    if (ret < 0) {
+        ret = -errno;
+        if ((errno != EADDRINUSE) || !silent_addinuse_err) {
+            vfs_error("bind(%s) failed: %m", g_sockaddr.sun_path);
+        }
+        goto out_close;
+    }
+
+    ret = listen(listen_fd, 128);
+    if (ret < 0) {
+        ret = -errno;
+        vfs_error("listen() failed: %m");
+        goto out_unlink;
+    }
+
+    vfs_error("closing socket, TODO implement listening for connections");
+    close(listen_fd);
+
+out_unlink:
+    vfs_unlink_socket(0);
+out_close:
+    close(listen_fd);
+out:
+    return ret;
+}
+
+/* return 0 or the (negative) value of errno in case of error */
+static int vfs_connect_and_act()
+{
+    ucs_vfs_sock_message_t vfs_msg_out;
+    int connfd;
+    int ret;
+
+    vfs_log("connecting to '%s'", g_sockaddr.sun_path);
+
+    connfd = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (connfd < 0) {
+        ret = -errno;
+        vfs_error("failed to create connection socket: %m");
+        return ret;
+    }
+
+    ret = connect(connfd, (const struct sockaddr*)&g_sockaddr,
+                  sizeof(g_sockaddr));
+    if (ret < 0) {
+        ret = -errno;
```

Comment:
set if right before `goto` operator and in other places too (up to you)
I guess it is better since can by some reason override the `ret` value in the future, it would lead to unexpected behavior
wdyt?

Reply:
i can't, afraid that vfs_log() might affect errno